iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,386 @@
+            Some(stream) => {
+                let (read, write) = stream.into_split();
+                Self {
+                    id,
+                    read: Some(read),
+                    write: Some(write),
+                    state,
+                    secret_key,
+                    public_key,
+                    cipher: None,
+                    network_addr: addr,
+                }
+            }
+        }
+    }
+
+    fn stream(&mut self) -> impl Stream<Item = MessageResult> + Send + 'static {
+        #[allow(clippy::expect_used)]
+        let mut read: OwnedReadHalf = self.read.take().expect("Unexpected");
+        stream! {
+            loop {
+                if let Err(e) = read.as_ref().readable().await {
+                    yield MessageResult::new_error(e);
+                    break;
+                }
+                let result = match read_message(&mut read).await {
+                    Ok(message) => MessageResult::new_message(message),
+                    Err(e) => MessageResult::new_error(e)
+                };
+
+                yield result;
+            }
+        }
+    }
+
+    async fn handshake(&mut self) -> Result<(), io::Error> {
+        match &self.state {
+            State::Connecting => {
+                // Connecting to peer
+                self.peer_connect().await
+            }
+            State::ConnectedTo => {
+                self.peer_send_hello().await?;
+            }
+            State::ConnectedFrom => {
+                self.peer_read_client_hello().await?;
+            }
+            State::Ready => {
+                warn!("Not doing handshake, already ready.");
+            }
+            State::Error => {
+                warn!("Not doing handshake in error state.");
+            }
+        }
+        Ok(())
+    }
+
+    // Reads client public key from client hello,
+    // creates shared secret and sends our public key to client
+    async fn peer_read_client_hello(&mut self) -> Result<(), io::Error> {
+        #[allow(clippy::expect_used)]
+        let read_half = self.read.as_mut().expect("Unexpected");
+        let public_key = read_client_hello(read_half).await?;
+        self.derive_shared_key(&public_key)?;
+        #[allow(clippy::expect_used)]
+        let mut write_half = self.write.as_mut().expect("Unexpected");
+        send_server_hello(&mut write_half, self.public_key.0.as_slice()).await?;
+        self.state = State::Ready;
+        Ok(())
+    }
+
+    /// Sends client hello with our public key
+    async fn peer_send_hello(&mut self) -> Result<(), io::Error> {
+        #[allow(clippy::expect_used)]
+        let mut write_half = self.write.as_mut().expect("Unexpected");
+        write_half.as_ref().writable().await?;
+        send_client_hello(&mut write_half, self.public_key.0.as_slice()).await?;
+        // Read server hello with node's public key
+        #[allow(clippy::expect_used)]
+        let read_half = self.read.as_mut().expect("Unexpected");
```

Comment:
There should be better messages for `expect`

Reply:
Technically, it is like `unwrap`, as we know that there is a value in that `Option`.